### ISSUE TITLE
[BUG] Keyboard actions don't work on a 2012 Panasonic Viera TV

### ISSUE THREAD

**Issue by RubenHetKonijn** [ttalvitie/browservice#34]
So I have this Panasonic Viera TV from 2012, I opened browservice and went to the server, I clicked on the search bar and attempted to type, No keyboard to type popped up (there was no text field) and I couldn't navigate anything. Connecting a keyboard to the TV also didn't work.

**Comment by ericmackrodt** [comment]
Browservice works by streaming down screenshots taken from the headless chrome instance, which means that there are no actual text fields.
It uses javascript to capture the keystrokes and send them back to the server, if the the field is focused on the server, those keystrokes will be typed in it.

Because of that, devices with browsers that pop a virtual keyboard up when you select a field won't work because you don't actually have a text field there.
The problem with the physical keyboard not working might be that the browser also ignores keystrokes if an actual text field is not selected.

You need to somehow verify if that's the case, if the browser can indeed capture the keystrokes in javascript without a field being selected. There might be a way to fix it on browservice if you do need an actual text field but that would require some trickery, it would be easier to make a change to support the TV browser if the browser can capture the keystrokes without the text field.

**Comment by RubenHetKonijn** [comment]
The TV browser / software is up to date and the browser does need an actual text field for it to work, what if someone made a transparent text field and put it over the browservice window, (and maybe only with a switch in the command) to allow this

**Comment by ericmackrodt** [comment]
Yes, that would be the kind of trickery that would need to be done to get it working on that specific browsers.

**Comment by limacohen** [comment]
I'm currently experiencing this but on an iPad 3 (MC705LE/A, iOS 9.3.5). Safari won't show browservice location bar, but I managed to see and interact with the controls using Dolphin browser. Sadly I can't type any address since the virtual keyboard won't come up and the physical one (Bluetooth) is ignored. I thought there might be some accessibility setting to keep the keyboard open, but if there is I can't find it.

So, +1 to the idea of the transparent text field for edge cases like these (that will become more common as time goes by and these touch/IoT devices get unsupported).

**Comment by ttalvitie** [comment]
It is indeed the case that the Browservice client does not currently have any support for opening virtual keyboards. In the end of https://github.com/ttalvitie/browservice/issues/16#issuecomment-782184340 I explain the situation more in detail. To recap, the server side cue for opening the virtual keyboard is easy to implement, but there does not seem to be an easy and portable way to force the virtual keyboard open on the client side. Thus we would probably need some less-than-optimal or non-portable hacks.

Supporting newer devices where virtual keyboards is not really my top focus in Browservice, but naturally I will try to help anyone who wants to take up the task. First it would be good to have some kind of POC on how to do this on the client side.

If there is great need for support on newer devices, it probably would make sense to implement a completely separate vice plugin for Browservice that would use more modern client web browser capabilities (such as HTML5 canvases and AJAX requests or even WebSockets) to get a smoother browsing experience; the current Retrojsvice plugin is highly inefficient because it has been designed with compatibility with very old browsers in mind. However, this is not on the top of my list, as I'm personally more interested in going to the other direction (implementing a vice plugin that enables usable native clients for even older systems, such as MS-DOS).

**Comment by limacohen** [comment]
Thank you for taking your time. I've read #16 as well now. I understand this would need a hackish fix that would be mostly irrelevant for older devices. Maybe years of being a webdev during pre-Chrome days have made me numb to per-OS/platform workarounds :D

In addition to the iPad 3 I have an eMac myself and I'm planning to use Browservice there too. But I started with this iOS device because if I can make the browser work, it still can fit nicely in my everyday workflow. I don't consider it "new" since it's almost 10yo but I see your point and I get that it's still in an uncanny valley of new-but-old tech.

Now regarding the actual problem, putting aside the virtual keyboard thing, the physical one also didn't work. That could probably mean that JS can't pick up the keystrokes unless they are focused on an actual text field. So, is there any major caveat in implementing a hidden text field and grabbing the keystrokes from there instead of the document/img? Maybe if we are lucky enough, focusing a 1x1px input can trigger the virtual keyboard as well, but I'm cool with compromising to a physical one.

**Comment by ttalvitie** [comment]
I just remembered that I actually have already implemented a hack like that where we have an invisible input element that we keep in focus; I needed this to avoid backspace going to the previous page on IE4. @RubenHetKonijn and @limacohen, could you check if you get the physical keyboard to work if you set `useBackspaceCaptureHack` to `true` on line 52 of [viceplugins/retrojsvice/html/main.html](https://github.com/ttalvitie/browservice/blob/v0.9.2.0/viceplugins/retrojsvice/html/main.html#L52), recompile with `make` and restart Browservice?

**Comment by RubenHetKonijn** [comment]
I'll need to reinstall Browservice for that to work and use WSL2, is this confirmed to work?

**Comment by ttalvitie** [comment]
WSL2 should work; I develop Browservice on WSL2.

**Comment by limacohen** [comment]
> could you check if you get the physical keyboard to work if you set `useBackspaceCaptureHack` to `true` on line 52 of [viceplugins/retrojsvice/html/main.html](https://github.com/ttalvitie/browservice/blob/v0.9.2.0/viceplugins/retrojsvice/html/main.html#L52), recompile with `make` and restart Browservice?

I could recompile it without issues as expected, but it didn't have any effect on the browser I previously tested (Dolphin). Safari still won't even load the GUI. I checked that the var was indeed changed in the source code and it was.

_**However**_ this time I did try with a very old version of Chrome, and it worked! Just the physical keyboard though. And it didn't have anything to do with the flag, I recompiled it back to default and behaves the same.

So, yay? :D

**Comment by limacohen** [comment]
While we are at it, I couldn't find in the docs if there's a way to pre-populate the (fake) address bar from the URL. That would save a lot of troubles.

**Comment by ttalvitie** [comment]
@limacohen OK, so for Dolphin my text field focus hack does not appear to work, and thus I'm out of ideas (as I can't experiment with that browser on my own). Still, it's good that you found one browser that Browservice works with

As for the question about prepopulating the address bar, you can use the `--start-page` command line option to specify the initial page. It is documented in command line option help accessed through `--help`.

**Comment by limacohen** [comment]
> @limacohen OK, so for Dolphin my text field focus hack does not appear to work, and thus I'm out of ideas (as I can't experiment with that browser on my own). Still, it's good that you found one browser that Browservice works with
I'm OK with that, I understand mine is and edge case. As I said before, maybe in the future more touch devices creep into the "old hardware" category and this becomes a more mainstream need.

> As for the question about prepopulating the address bar, you can use the `--start-page` command line option to specify the initial page. It is documented in command line option help accessed through `--help`.

**Comment by limacohen** [comment]
> @limacohen OK, so for Dolphin my text field focus hack does not appear to work, and thus I'm out of ideas (as I can't experiment with that browser on my own). Still, it's good that you found one browser that Browservice works with

I'm OK with that, I understand mine is an edge case. As I said before, maybe in the future more touch devices creep into the "old hardware" category and this becomes a more mainstream need.

> As for the question about prepopulating the address bar, you can use the `--start-page` command line option to specify the initial page. It is documented in command line option help accessed through `--help`.

Thank you, but I was thinking more in the lines of the GUI browser client, so I can change the address in the (real) location bar using any keyboard (virtual or physical) and change the current loaded page. IMO this would be an elegant workaround for this particular issue, and serve as a feature for creating shortcuts to different pages without the need to restart the server.
